public/app/features/browse-dashboards/components/DashboardsTree.tsx

```diff
@@ -0,0 +1,155 @@
+import { css } from '@emotion/css';
+import React, { useMemo } from 'react';
+import { CellProps, Column, TableInstance, useTable } from 'react-table';
+import { FixedSizeList as List } from 'react-window';
+
+import { GrafanaTheme2 } from '@grafana/data';
+import { useStyles2 } from '@grafana/ui';
+
+import { DashboardsTreeItem, INDENT_AMOUNT_CSS_VAR } from '../types';
+
+import { NameCell } from './NameCell';
+import { TypeCell } from './TypeCell';
+
+interface DashboardsTreeProps {
+  items: DashboardsTreeItem[];
+  width: number;
+  height: number;
+  onFolderClick: (uid: string, newOpenState: boolean) => void;
+}
+
+type DashboardsTreeColumn = Column<DashboardsTreeItem>;
+
+const HEADER_HEIGHT = 35;
+const ITEM_HEIGHT = 35;
+
+export function DashboardsTree({ items, width, height, onFolderClick }: DashboardsTreeProps) {
+  const styles = useStyles2(getStyles);
+
+  const tableColumns = useMemo(() => {
+    const checkboxColumn: DashboardsTreeColumn = {
+      id: 'checkbox',
+      Header: () => <input type="checkbox" />,
+      Cell: () => <input type="checkbox" />,
+    };
+
+    const nameColumn: DashboardsTreeColumn = {
+      id: 'name',
+      Header: <span style={{ paddingLeft: 20 }}>Name</span>,
+      Cell: (props: CellProps<DashboardsTreeItem, unknown>) => <NameCell {...props} onFolderClick={onFolderClick} />,
+    };
+
+    const typeColumn: DashboardsTreeColumn = {
+      id: 'type',
+      Header: 'Type',
+      Cell: TypeCell,
+    };
+
+    return [checkboxColumn, nameColumn, typeColumn];
+  }, [onFolderClick]);
+
+  const table = useTable({ columns: tableColumns, data: items });
+  const { getTableProps, getTableBodyProps, headerGroups } = table;
+
+  return (
+    <div {...getTableProps()} className={styles.tableRoot} role="table">
+      {headerGroups.map((headerGroup) => {
+        const { key, ...headerGroupProps } = headerGroup.getHeaderGroupProps({
+          style: { width },
+        });
+
+        return (
+          <div key={key} {...headerGroupProps} className={styles.headerRow}>
+            {headerGroup.headers.map((column) => {
+              const { key, ...headerProps } = column.getHeaderProps();
+
+              return (
+                <div key={key} {...headerProps} role="columnheader" className={styles.cell}>
+                  {column.render('Header')}
+                </div>
+              );
```

Comment:
Maybe these can be extracted to separate functions for better readability?

Reply:
I've been thinking the same thing, so I gave it a try but I don't think it aided much. I think one of the main problems is how eslint needs us to explicitly set key on the map children - it cannot see that getHeaderProps returns key in its type, so we still need to do the same destructure and spread, which I think is the main part that makes this tricky to read.

Reply:
If people feel strongly about this happy to take another pass.